internal/batch/batch_processor.go

```diff
@@ -101,103 +130,249 @@ func (bw *batchWork) estimateSize() int64 {
 	return sizeEstimate
 }
 
-// The assemblyLoop accepts work into the pipe as quickly as possible.
-// It dispatches work asynchronously to the persistenceLoop, which is responsible for
-// calling back each piece of work once persisted into a batch
-// (doesn't wait until that batch is sealed/dispatched).
-// The assemblyLoop seals batches when they are full, or timeout.
+func (bp *batchProcessor) status() *ProcessorStatus {
+	bp.statusMux.Lock()
+	defer bp.statusMux.Unlock()
+	return &ProcessorStatus{
+		Dispatcher: bp.conf.dispatcherName,
+		Name:       bp.conf.name,
+		Status:     bp.flushStatus, // copy
+	}
+}
+
+func (bp *batchProcessor) newAssembly(initalWork ...*batchWork) {
+	bp.assemblyID = fftypes.NewUUID()
+	bp.assemblyQueue = append([]*batchWork{}, initalWork...)
+	bp.assemblyQueueBytes = batchSizeEstimateBase
+}
+
+// addWork adds the work to the assemblyQueue, and calculates if we have overflowed with this work.
+// We check for duplicates, and add the work in sequence order.
+// This helps in the case for parallel REST APIs all committing to the DB at a similar time.
+// With a sufficient batch size and batch timeout, the batch will still dispatch the messages
+// in DB sequence order (although this is not guaranteed).
+func (bp *batchProcessor) addWork(newWork *batchWork) (full, overflow bool) {
+	newQueue := make([]*batchWork, 0, len(bp.assemblyQueue)+1)
+	added := false
+	skip := false
+	// Check it's not in the recently flushed lish
+	for _, flushedSequence := range bp.flushedSequences {
+		if newWork.msg.Sequence == flushedSequence {
+			log.L(bp.ctx).Debugf("Ignoring add of recently flushed message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			skip = true
+			break
+		}
+	}
+	// Build the new sorted work list, checking there for duplicates too
+	for _, work := range bp.assemblyQueue {
+		if newWork.msg.Sequence == work.msg.Sequence {
+			log.L(bp.ctx).Debugf("Ignoring duplicate add of message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			skip = true
+		}
+		if !added && !skip && newWork.msg.Sequence < work.msg.Sequence {
+			newQueue = append(newQueue, newWork)
+			added = true
+		}
+		newQueue = append(newQueue, work)
+	}
```

Comment:
It also feels like much of this loop could be replaced by a single `append` of the new work and then a call to `sort.Slice()`. Would still need a check for duplicates ahead of that though, so unsure of the ultimate performance impact. Just a thought.

Reply:
So I think this ends up as more code in the end of the day. We have to have all the branches that we have today, but _also_ create a custom function do pass into the `sort.Slice()` to do the comparison.

As for whether it's more performant in practice to do a single traversal and allocate memory for a new pointer array each time, vs only re-allocate the array when we add and do a sort, I'm not sure.

My gut was re-allocating the array each time and doing a single pass optimizes for the case where we're adding work - which is the most common case.

Reply:
Yea, I think with the other optimizations added to abort early in the case of a duplicate, this now reads pretty clean and doesn't need any further tweaking.